src/main/java/org/opensearch/search/relevance/transformer/personalizeintelligentranking/reranker/impl/AmazonPersonalizedRankerImpl.java

```diff
@@ -82,14 +88,82 @@ public SearchHits rerank(SearchHits hits, PersonalizeRequestParameters requestPa
+            }
+
+            // Build a map with key as item id and value as open search scores and another map
+            // with key as item id and value as corresponding search hit
+            for (SearchHit hit : originalHits) {
+                if (!itemIdfield.isEmpty()){
+                    idToOpenSearchScoreMap.put(hit.getSourceAsMap().get(itemIdfield).toString(), hit.getScore());
+                    itemIdToSearchHitMap.put(hit.getSourceAsMap().get(itemIdfield).toString(), hit);
+                }
+                else{
+                    idToOpenSearchScoreMap.put(hit.getId(), hit.getScore());
+                    itemIdToSearchHitMap.put(hit.getId(), hit);
+                }
+            }
+
+
+            float weight = (float) rankerConfig.getWeight();
+            SearchHits newHits = combineScores(idToPersonalizeRankingScoreMap, idToOpenSearchScoreMap,
+                    itemIdToSearchHitMap, hits.getTotalHits(), weight);
+            return newHits;
         } catch (Exception ex) {
             logger.error("Failed to re rank with Personalize. Returning original search results without Personalize re ranking.", ex);
             return hits;
         }
     }
 
+    //Combine open search hits and personalize campaign response
+    public SearchHits combineScores(Map<String, Float> idToPersonalizeRankingScoreMap,
+                                     Map<String, Float> idToOpenSearchScoreMap,
+                                     Map<String, SearchHit> itemIdToSearchHitMap,
+                                     TotalHits totalHits, float weight) {
+        //Update open search score based on the personalize campaign response for each item id
+        List<String> openSearchItemId = new ArrayList<String>(idToOpenSearchScoreMap.keySet());
+        for (String itemId : openSearchItemId) {
+            if(idToPersonalizeRankingScoreMap.containsKey(itemId)){
+                float personalizedScore = idToPersonalizeRankingScoreMap.get(itemId);
+                float openSearchScore = idToOpenSearchScoreMap.get(itemId);
+                float combinedScore = (float) (weight / Math.log(openSearchScore + 1)
+                        + (1 - weight) / Math.log(personalizedScore + 1));
+                idToOpenSearchScoreMap.put(itemId, combinedScore);
+            }
+        }
+
+        //Create a new list of search hits in the decreasing order of the combined scores
+        Map<String, Float> sortedScores = sortByValue(idToOpenSearchScoreMap);
+
+        List<SearchHit> rerankedHits = sortedScores.keySet().stream()
+                .map(itemId -> {
+                    SearchHit hit = itemIdToSearchHitMap.get(itemId);
+                    hit.score(sortedScores.get(itemId));
+                    return hit;
+                })
+                .collect(Collectors.toList());
+        float maxScore = sortedScores.values().stream().max(Float::compare).orElse(0f);
```

Comment:
You're iterating through `sortedScores` twice here.

Also all these maps are confusing. The only one worth keeping (IMO) is `idToPersonalizeRankingScoreMap`.

I would do something like the following logic:
```

List<SearchHit> rerankedHits = new ArrayList<>(originalHits.length);
float maxScore = 0f;
for (SearchHit hit : originalHits) {
  float openSearchScore = hit.getScore();
  float personalizedScore = idToPersonalizeRankingScoreMap.get(hit.getId());
  float combinedScore = (float) (weight / Math.log(openSearchScore + 1)
                        + (1 - weight) / Math.log(personalizedScore + 1));
  maxScore = Math.max(maxScore, combinedScore);
  hit.score(combinedScore);
  rerankedHits.add(hit);
}
rerankedHits.sort(Comparator.comparing(h -> h.getScore()).reverse());
```

That would eliminate the need for two of the maps and the `sortByValue` method.

Reply:
Discussed w/ @mishprs -- this change will go into the next PR.